extmod/uasyncio/core.py

```diff
@@ -145,81 +146,88 @@ def create_task(coro):
     _task_queue.push(t)
     return t
 
-
 # Keep scheduling tasks until there are none left to schedule
 def run_until_complete(main_task=None):
     global cur_task
     excs_all = (CancelledError, Exception)  # To prevent heap allocation in loop
     excs_stop = (CancelledError, StopIteration)  # To prevent heap allocation in loop
-    while True:
-        # Wait until the head of _task_queue is ready to run
-        dt = 1
-        while dt > 0:
-            dt = -1
+    try:
+        _io_queue.wait_io_event(0)
+        while True:
+            # Wait until the head of _task_queue is ready to run
             t = _task_queue.peek()
             if t:
                 # A task waiting on _task_queue; "ph_key" is time to schedule task at
-                dt = max(0, ticks_diff(t.ph_key, ticks()))
+                dt = ticks_diff(t.ph_key, ticks())
+                if dt > 0:
+                    _io_queue.wait_io_event(dt)
```

Comment:
I'm not sure this is correct, because it may be possible to starve tasks waiting on an IO event.  Eg if there are two tasks that keep rescheduling themselves immediately then we always get `dt == 0` here and `wait_io_event()` is never called.  So tasks that are on the IO queue will remain there forever, even if they become readable/writable.

Reply:
If what I say above is true then we should write a test for this case, see it fail, then add that test to the existing test suite to catch such cases in the future.

Reply:
Good catch, @dpgeorge. I changed it to always call `wait_io_event` but still avoid the call to `max()`